File: app/schemas/eval.py
from pydantic import BaseModel, Field
from typing import Literal, Optional

# --- Test Data Management Schemas ---
class EvalDataItem(BaseModel):
    index: int
    input: str
    expected_output: str
    source: Literal["llm", "human"]
    reviewed: bool

class EvalDataUpdateRequest(BaseModel):
    input: Optional[str] = None
    expected_output: Optional[str] = None
    reviewed: Optional[bool] = None
    model_config = {"extra": "forbid"}

class EvalDataGenerateRequest(BaseModel):
    num_cases: int = Field(default=5, ge=1, le=20)
    model_config = {"extra": "forbid"}

class EvalDataGenerateResponse(BaseModel):
    generated_count: int
    total_count: int
    file_path: str

# --- Evaluation Request & Response Schemas ---
class DeterministicEvalRequest(BaseModel):
    metrics: list[str] = ["correctness"]
    model_config = {"extra": "forbid"}

class NonDeterministicEvalRequest(BaseModel):
    inputs: list[str] = Field(
        ..., 
        description="Ad-hoc user queries to test agent behavior and tool usage",
        min_length=1
    )
    metrics: list[str] = ["task_completion"]
    model_config = {"extra": "forbid"}

class EvalMetricResult(BaseModel):
    metric_name: str
    score: float
    threshold: float
    passed: bool
    reason: str | None = None

class EvalCaseResult(BaseModel):
    agent_input: str
    agent_output: str
    expected_output: str | None = None
    metric_results: list[EvalMetricResult]

    @property
    def all_passed(self) -> bool:
        return all(m.passed for m in self.metric_results)

class EvalReportResponse(BaseModel):
    agent_name: str
    evaluation_type: Literal["deterministic", "non_deterministic"]
    total_cases: int
    all_passed: bool
    results: list[EvalCaseResult]

File: frontend/src/types/eval.ts
export type EvalType = 'non_deterministic' | 'deterministic';

export type DeterministicMetricKey = 'correctness' | 'tool_correctness';
export type NonDeterministicMetricKey = 'step_efficiency' | 'task_completion' | 'hallucination';

export type MetricKey = DeterministicMetricKey | NonDeterministicMetricKey;

export interface MetricDefinition {
  key: MetricKey;
  label: string;
  description: string;
  defaultThreshold: number;
  evalType: EvalType;
}

export const METRIC_DEFINITIONS: Record<MetricKey, MetricDefinition> = {
  step_efficiency: {
    key: 'step_efficiency',
    label: 'Step Efficiency',
    description: 'Evaluates if the agent completed the task with minimal LLM calls and tools without looping or redundancy.',
    defaultThreshold: 0.5,
    evalType: 'non_deterministic',
  },
  task_completion: {
    key: 'task_completion',
    label: 'Task Completion',
    description: 'Verifies whether the agent fully satisfied all instructions and goals in the user prompt.',
    defaultThreshold: 0.7,
    evalType: 'non_deterministic',
  },
  hallucination: {
    key: 'hallucination',
    label: 'Hallucination Check',
    description: 'Checks if agent statements are factually grounded in context without fabricating information.',
    defaultThreshold: 0.5,
    evalType: 'non_deterministic',
  },
  correctness: {
    key: 'correctness',
    label: 'Output Correctness',
    description: 'Evaluates semantic and factual alignment of the output against the expected golden answer.',
    defaultThreshold: 0.3,
    evalType: 'deterministic',
  },
  tool_correctness: {
    key: 'tool_correctness',
    label: 'Tool Correctness',
    description: 'Validates that the exact tools expected were called with appropriate parameters.',
    defaultThreshold: 0.7,
    evalType: 'deterministic',
  },
};

export interface DeterministicEvalRequest {
  metrics: string[];
}

export interface NonDeterministicEvalRequest {
  inputs: string[];
  metrics: string[];
}

export interface EvalMetricResult {
  metric_name: string;
  score: number;
  threshold: number;
  passed: boolean;
  reason?: string | null;
}

export interface EvalCaseResult {
  agent_input: string;
  agent_output: string;
  expected_output?: string | null;
  metric_results: EvalMetricResult[];
}

export interface EvalReportResponse {
  agent_name: string;
  evaluation_type: EvalType;
  total_cases: number;
  all_passed: boolean;
  results: EvalCaseResult[];
}
